File: anthropic/java/anthropic/ClusterMode2.java
package anthropic;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/** Distributed median (from cluster_mode2.py). */
public class ClusterMode2 {

    static class Worker {
        final int workerId;
        final int k;
        final List<Integer> localData;
        final ClusterMode2 cluster;

        Worker(int workerId, int k, List<Integer> localData, ClusterMode2 cluster) {
            this.workerId = workerId;
            this.k = k;
            this.localData = localData;
            this.cluster = cluster;
        }

        void sendAsyncMessage(int targetWorkerId, String payload) {
            cluster.sendAsyncMessage(targetWorkerId, payload);
        }

        String receive() {
            return cluster.receive(workerId);
        }

        void sendLocalStats() {
            if (!localData.isEmpty()) {
                int mn = Integer.MAX_VALUE;
                int mx = Integer.MIN_VALUE;
                for (int v : localData) {
                    mn = Math.min(mn, v);
                    mx = Math.max(mx, v);
                }
                sendAsyncMessage(0, "STAT " + localData.size() + " " + mn + " " + mx);
            }
            sendAsyncMessage(0, "DONE_STATS");
        }

        int[] aggregateGlobalStats() {
            int total = 0;
            int gLo = Integer.MAX_VALUE;
            int gHi = Integer.MIN_VALUE;
            int done = 0;
            while (done < k) {
                String msg = receive();
                if (msg.startsWith("STAT")) {
                    String[] p = msg.split(" ");
                    int s = Integer.parseInt(p[1]);
                    total += s;
                    if (s > 0) {
                        gLo = Math.min(gLo, Integer.parseInt(p[2]));
                        gHi = Math.max(gHi, Integer.parseInt(p[3]));
                    }
                } else if ("DONE_STATS".equals(msg)) {
                    done++;
                }
            }
            if (total == 0) return new int[] {0, 0, 0};
            return new int[] {total, gLo, gHi};
        }

        void sendCountsForPivot(int pivot) {
            int le = 0;
            for (int v : localData) {
                if (v <= pivot) le++;
            }
            sendAsyncMessage(0, "CNT " + le);
            sendAsyncMessage(0, "DONE_CNT");
        }

        int aggregateLeCount() {
            int total = 0;
            int done = 0;
            while (done < k) {
                String msg = receive();
                if (msg.startsWith("CNT")) {
                    total += Integer.parseInt(msg.split(" ")[1]);
                } else if ("DONE_CNT".equals(msg)) {
                    done++;
                }
            }
            return total;
        }
    }

    final int k;
    final List<List<Integer>> shards = new ArrayList<>();
    final List<List<String>> mailboxes = new ArrayList<>();
    final List<Integer> readIndices = new ArrayList<>();
    final Worker[] workers;

    public ClusterMode2(List<Integer> data, int k) {
        this.k = k;
        for (int i = 0; i < k; i++) {
            shards.add(new ArrayList<>());
            mailboxes.add(new ArrayList<>());
            readIndices.add(0);
        }
        int n = data.size();
        int base = n / k;
        int rem = n % k;
        int idx = 0;
        for (int w = 0; w < k; w++) {
            int cs = base + (w < rem ? 1 : 0);
            for (int j = 0; j < cs; j++) {
                shards.get(w).add(data.get(idx++));
            }
        }
        workers = new Worker[k];
        for (int i = 0; i < k; i++) {
            workers[i] = new Worker(i, k, shards.get(i), this);
        }
    }

    void sendAsyncMessage(int targetWorkerId, String payload) {
        mailboxes.get(targetWorkerId).add(payload);
    }

    String receive(int workerId) {
        List<String> box = mailboxes.get(workerId);
        int ri = readIndices.get(workerId);
        if (ri < box.size()) {
            readIndices.set(workerId, ri + 1);
            return box.get(ri);
        }
        return "";
    }

    public double findMedian() {
        for (Worker w : workers) {
            w.sendLocalStats();
        }
        int[] g = workers[0].aggregateGlobalStats();
        int n = g[0];
        if (n == 0) return Double.NaN;
        int lo = g[1];
        int hi = g[2];
        int i = (n - 1) / 2;
        int j = n / 2;
        return (selectKth(i, lo, hi) + selectKth(j, lo, hi)) / 2.0;
    }

    int selectKth(int targetIndex, int low, int high) {
        int left = low;
        int right = high;
        while (left < right) {
            int mid = left + (right - left) / 2;
            for (Worker w : workers) {
                w.sendCountsForPivot(mid);
            }
            int le = workers[0].aggregateLeCount();
            if (le <= targetIndex) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }

    public static void main(String[] args) {
        List<Integer> d = Arrays.asList(1, 3, 5);
        System.out.println(new ClusterMode2(new ArrayList<>(d), 2).findMedian());
    }
}

File: anthropic/python/cluster_mode2.py
import sys
from typing import List

class Worker:
    def __init__(self, workerId: int, k: int, localData: List[int], cluster):
        self.workerId = workerId
        self.k = k
        self.localData = localData
        self.cluster = cluster

    # Sends a string-represented payload data to the specified worker
    # asynchronously. You should NOT modify this method.
    def sendAsyncMessage(self, targetWorkerId: int, payload: str):
        self.cluster.sendAsyncMessage(targetWorkerId, payload)

    # Receives a string-represented payload data from the worker's mailbox. You
    # should NOT modify this method.
    def receive(self) -> str:
        return self.cluster.receive(self.workerId)

    def sendLocalStats(self) -> None:
        d = self.localData
        if d:
            self.sendAsyncMessage(0, f"STAT {len(d)} {min(d)} {max(d)}")
        self.sendAsyncMessage(0, "DONE_STATS")

    def aggregateGlobalStats(self) -> List[int]:
        total = 0
        g_lo, g_hi = sys.maxsize, -sys.maxsize - 1
        done = 0
        while done < self.k:
            msg = self.cluster.receive(self.workerId)
            if msg.startswith("STAT"):
                _, s, lo, hi = msg.split()
                s = int(s)
                total += s
                if s > 0:
                    g_lo = min(g_lo, int(lo))
                    g_hi = max(g_hi, int(hi))
            elif msg == "DONE_STATS":
                done += 1
        if total == 0:
            return [0, 0, 0]
        return [total, g_lo, g_hi]

    def sendCountsForPivot(self, pivot: int) -> None:
        # #{x : x <= pivot}; coordinator finds smallest v with global_le(v) > targetIndex
        le_c = sum(1 for v in self.localData if v <= pivot)
        self.sendAsyncMessage(0, f"CNT {le_c}")
        self.sendAsyncMessage(0, "DONE_CNT")

    def aggregateLeCount(self) -> int:
        total = 0
        done = 0
        while done < self.k:
            msg = self.cluster.receive(self.workerId)
            if msg.startswith("CNT"):
                _, a = msg.split()
                total += int(a)
            elif msg == "DONE_CNT":
                done += 1
        return total


class Cluster:
    def __init__(self, data: List[int], k: int):
        self.k = k

        self.shards = []
        for i in range(k):
            self.shards.append([])

        totalSize = len(data)
        baseSize = totalSize // k
        remainder = totalSize % k

        index = 0
        for w in range(k):
            chunkSize = baseSize
            if w < remainder:
                chunkSize = chunkSize + 1
            for j in range(chunkSize):
                self.shards[w].append(data[index])
                index = index + 1

        self.mailboxes = {}
        self.readIndices = {}
        for i in range(k):
            self.mailboxes[i] = []
            self.readIndices[i] = 0

        self.workers = [None] * k
        for i in range(k):
            self.workers[i] = Worker(i, k, self.shards[i], self)

    def sendAsyncMessage(self, targetWorkerId: int, payload: str):
        self.mailboxes[targetWorkerId].append(payload)

    def receive(self, workerId: int) -> str:
        myMailbox = self.mailboxes[workerId]
        idx = self.readIndices[workerId]
        if idx < len(myMailbox):
            self.readIndices[workerId] = idx + 1
            return myMailbox[idx]

        return ""

    def findMedian(self) -> float:
        for w in self.workers:
            w.sendLocalStats()
        n, lo, hi = self.workers[0].aggregateGlobalStats()
        if n == 0:
            return float("nan")
        i, j = (n - 1) // 2, n // 2
        return (self.selectKth(i, lo, hi) + self.selectKth(j, lo, hi)) / 2.0

    def selectKth(self, targetIndex: int, low: int, high: int) -> int:
        # 0-based targetIndex-th smallest = smallest v with #{x <= v} > targetIndex
        left, right = low, high
        while left < right:
            mid = left + (right - left) // 2
            for w in self.workers:
                w.sendCountsForPivot(mid)
            le = self.workers[0].aggregateLeCount()
            if le <= targetIndex:
                left = mid + 1
            else:
                right = mid
        return left

    @staticmethod
    def main():
        Cluster.test1()
        Cluster.test2()
        Cluster.test3()
        Cluster.test4()

    @staticmethod
    def test1():
        print("===== Test 1 =====")
        data = [1, 3, 5]
        cluster = Cluster(data, 2)
        print(cluster.findMedian())  # Expected: 3.0

    @staticmethod
    def test2():
        print("===== Test 2 =====")
        data = [1, 2, 3, 4]
        cluster = Cluster(data, 3)
        print(cluster.findMedian())  # Expected: 2.5

    @staticmethod
    def test3():
        print("===== Test 3 =====")
        data = []
        for i in range(50):
            data.append(10)
        for i in range(30):
            data.append(20)
        for i in range(20):
            data.append(-5)
        cluster = Cluster(data, 5)
        print(cluster.findMedian())  # Expected: 10.0

    @staticmethod
    def test4():
        print("===== Test 4 =====")
        data = [10, 20, 30]
        cluster = Cluster(data, 5)
        print(cluster.findMedian())  # Expected: 20.0


if __name__ == "__main__":
    Cluster.main()